controller_manager/src/controller_manager.cpp

```diff
@@ -2030,30 +2030,35 @@ controller_interface::return_type ControllerManager::update(
   ++update_loop_counter_;
   update_loop_counter_ %= update_rate_;
 
-  for (auto loaded_controller : rt_controller_list)
+  for (const auto & loaded_controller : rt_controller_list)
   {
     // TODO(v-lopez) we could cache this information
     // https://github.com/ros-controls/ros2_control/issues/153
     if (!loaded_controller.c->is_async() && is_controller_active(*loaded_controller.c))
     {
       const auto controller_update_rate = loaded_controller.c->get_update_rate();
-      const auto controller_update_factor =
-        (controller_update_rate == 0) || (controller_update_rate >= update_rate_)
-          ? 1u
-          : update_rate_ / controller_update_rate;
+      const bool run_controller_at_cm_rate =
+        (controller_update_rate == 0) || (controller_update_rate >= update_rate_);
+      const auto controller_period =
+        run_controller_at_cm_rate ? period
+                                  : rclcpp::Duration::from_seconds((1.0 / controller_update_rate));
+
+      bool controller_go = (time == rclcpp::Time(0)) ||
+                           (time.seconds() >= loaded_controller.next_update_cycle_time->seconds());
 
-      bool controller_go = ((update_loop_counter_ % controller_update_factor) == 0);
       RCLCPP_DEBUG(
         get_logger(), "update_loop_counter: '%d ' controller_go: '%s ' controller_name: '%s '",
         update_loop_counter_, controller_go ? "True" : "False",
         loaded_controller.info.name.c_str());
 
       if (controller_go)
       {
-        auto controller_ret = loaded_controller.c->update(
-          time, (controller_update_factor != 1u)
-                  ? rclcpp::Duration::from_seconds(1.0 / controller_update_rate)
-                  : period);
+        auto controller_ret =
+          loaded_controller.c->update(time, run_controller_at_cm_rate ? period : controller_period);
+
+        if (*loaded_controller.next_update_cycle_time == rclcpp::Time(0))
+          *loaded_controller.next_update_cycle_time = time;
+        *loaded_controller.next_update_cycle_time += controller_period;
```

Comment:
This is hard do read because of the formatting. Should it be this:

```suggestion
        if (*loaded_controller.next_update_cycle_time == rclcpp::Time(0))
        {
          *loaded_controller.next_update_cycle_time = time;
        }
        *loaded_controller.next_update_cycle_time += controller_period;
```

Reply:
Sure. Makes sense